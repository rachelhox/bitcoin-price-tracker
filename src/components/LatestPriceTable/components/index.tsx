import PriceRow from './PriceRow';

export { PriceRow };
