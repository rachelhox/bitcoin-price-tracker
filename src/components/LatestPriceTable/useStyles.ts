import { css, useTheme } from '@mui/material';

const useStyles = () => {
  const theme = useTheme();
  return {
    root: css({
      padding: theme.spacing(0, 3),
      display: 'flex',
      flexDirection: 'column',
      alignItems: 'center',
      [theme.breakpoints.down('laptop')]: {
        minHeight: 'auto',
      },
      [theme.breakpoints.up('laptop')]: {
        height: '100vh',
      },
    }),
    maxWidth: css({
      width: '100%',
      paddingBottom: theme.spacing(4),
      [theme.breakpoints.up('tablet')]: {
        maxWidth: '1200px',
        width: '80%',
        paddingBottom: 0,
      },
      [theme.breakpoints.up('laptop')]: {
        paddingBottom: theme.spacing(4),
      },
    }),
    container: css({
      display: 'flex',
      justifyContent: 'space-between',
      alignItems: 'center',
      flexDirection: 'column',
      paddingTop: theme.spacing(15),
      [theme.breakpoints.up('laptop')]: {
        flexDirection: 'row',
      },
    }),
    h3: css({
      fontSize: theme.spacing(1.5),
      color: theme.palette.common.white,
      width: '100%',
      textAlign: 'left',
      [theme.breakpoints.up('laptop')]: {
        padding: theme.spacing(0, 0, 2, 0),
        fontSize: theme.spacing(2),
      },
    }),
  };
};

export default useStyles;
