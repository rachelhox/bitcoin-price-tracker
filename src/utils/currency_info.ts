interface IStringIndex extends Record<string, any> {}

/**
 * Takes the key name and returns the location. Will return null if undefined
 */

const currencies: IStringIndex = {
  usd: {
    name: 'US Dollar',
    key: 'usd',
    unit: '$',
    type: 'fiat',
  },
  aed: {
    name: 'United Arab Emirates Dirham',
    key: 'aed',
    unit: 'DH',
    type: 'fiat',
  },
  eur: {
    name: 'Euro',
    key: 'eur',
    unit: '€',
    type: 'fiat',
  },
  aud: {
    name: 'Australian Dollar',
    key: 'aud',
    unit: 'A$',
    type: 'fiat',
  },
  cad: {
    name: 'Canadian Dollar',
    key: 'cad',
    unit: 'CA$',
    type: 'fiat',
  },
  chf: {
    name: 'Swiss Franc',
    key: 'chf',
    unit: 'Fr.',
    type: 'fiat',
  },
  gbp: {
    name: 'British Pound Sterling',
    key: 'gbp',
    unit: '£',
    type: 'fiat',
  },
  hkd: {
    name: 'Hong Kong Dollar',
    key: 'hkd',
    unit: 'HK$',
    type: 'fiat',
  },
  sgd: {
    name: 'Singapore Dollar',
    key: 'sgd',
    unit: 'S$',
    type: 'fiat',
  },
  cny: {
    name: 'Chinese Yuan',
    key: 'cny',
    unit: '¥',
    type: 'fiat',
  },
  jpy: {
    name: 'Japanese Yen',
    key: 'jpy',
    unit: '¥',
    type: 'fiat',
  },
  inr: {
    name: 'Indian Rupee',
    key: 'inr',
    unit: '₹',
    type: 'fiat',
  },
  krw: {
    name: 'South Korean Won',
    key: 'krw',
    unit: '₩',
    type: 'fiat',
  },
};

export const getCurrencyInfo = (key: string | number) => {
  return currencies[key] ?? {};
};
