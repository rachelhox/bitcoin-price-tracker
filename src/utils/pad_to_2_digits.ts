/* eslint-disable */

export const padTo2Digits = (num: any) => {
  return String(num).padStart(2, '0');
};
