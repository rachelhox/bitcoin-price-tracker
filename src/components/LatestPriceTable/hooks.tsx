/* eslint-disable no-undef */
/* eslint-disable array-callback-return */
/* eslint-disable no-console */
import { useState, useEffect } from 'react';
import axios from 'axios';
import useSWR from 'swr';
import { getCurrencyInfo } from '@src/utils';
import { currencyKeys } from './config';

export const usePriceTableHook = () => {
  const currencies = currencyKeys.map((x: string) => getCurrencyInfo(x));
  const [priceTableData, setPriceTableData] = useState<
    Array<CurrencyValueProp>
  >([]);
  const fetcher = (url: string) => axios.get(url).then((res) => res.data);
  // eslint-disable-next-line no-unused-vars
  const { data: ratesData, isValidating } = useSWR(
    `${process.env.NEXT_PUBLIC_COINGECKO_API_URL}/exchange_rates`,
    fetcher,
    { refreshInterval: 60000 }
  );
  useEffect(() => {
    setPriceTableData([]);
    if (ratesData) {
      const { rates } = ratesData;
      currencies.map((currency: Currency) => {
        const currencyTarget = rates[currency.key];
        setPriceTableData((prev) => [
          ...prev,
          {
            ...currency,
            value: currencyTarget.value,
          },
        ]);
      });
    }
  }, [ratesData, isValidating]);

  return { priceTableData };
};
