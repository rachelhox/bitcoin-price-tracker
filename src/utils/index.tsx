import { getCurrencyInfo } from './currency_info';
import { convertToMoney } from './convert_to_money';
import { padTo2Digits } from './pad_to_2_digits';

export { getCurrencyInfo, convertToMoney, padTo2Digits };
