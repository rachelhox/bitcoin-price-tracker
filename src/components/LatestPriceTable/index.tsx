/* eslint-disable no-unused-vars */
import { Box, Typography } from '@mui/material';
import { TimePeriodButton } from '@components';
import useStyles from './useStyles';
import { PriceRow } from './components';
import { usePriceTableHook } from './hooks';

const LatestPriceTable = () => {
  const styles = useStyles();
  const { priceTableData } = usePriceTableHook();

  return (
    <Box css={styles.root}>
      <Box css={styles.maxWidth}>
        <Box css={styles.container}>
          <Typography variant="h1" css={styles.h1}>
            BTC exchange rates
          </Typography>
        </Box>
        <PriceRow priceTableData={priceTableData} />
      </Box>
    </Box>
  );
};

export default LatestPriceTable;
