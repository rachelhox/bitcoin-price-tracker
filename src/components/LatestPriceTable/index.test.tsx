import renderer from 'react-test-renderer';
import { MockTheme } from '@tests/utils';
import LatestPriceTable from '.';

// ==================================
// mocks
// ==================================

jest.mock('./components', () => ({
  PriceRow: (props: any) => <div id="PriceRow" {...props} />,
}));

// ==================================
// unit tests
// ==================================
describe('component: LatestPriceTable', () => {
  it('matches snapshot', () => {
    const component = renderer.create(
      <MockTheme>
        <LatestPriceTable />
      </MockTheme>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
});
