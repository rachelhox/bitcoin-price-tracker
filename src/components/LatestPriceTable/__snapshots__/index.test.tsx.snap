// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: LatestPriceTable matches snapshot 1`] = `
<div
  className="MuiBox-root css-1bg7lxm-LatestPriceTable"
>
  <div
    className="MuiBox-root css-1ov9l26-LatestPriceTable"
  >
    <div
      className="MuiBox-root css-adzkba-LatestPriceTable"
    >
      <h3
        className="MuiTypography-root MuiTypography-h3 css-8ipx91-MuiTypography-root-LatestPriceTable"
      >
        *BTC exchange rates update every 1 minute
      </h3>
    </div>
    <div
      id="PriceRow"
      priceTableData={[]}
    />
  </div>
</div>
`;
